graphql/schemabuilder/schema.go

```diff
@@ -227,6 +272,60 @@ func (s *Schema) Build() (*graphql.Schema, error) {
 		}
 
 		sb.objects[typ] = object
+
+		// Don't enforce type name uniqueness for objects used for introspection
+		// because they are not included in the schema.
+		if len(object.Name) >= 2 && object.Name[:2] == "__" {
+			continue
+		}
+
+		for methodName, method := range object.Methods {
+			// Don't enforce type name uniqueness for methods defined on objects
+			// used for introspection because they are not included in the schema.
+			if len(methodName) > 2 && methodName[:2] == "__" {
+				continue
+			}
+
+			// It is valid for federated fieldfuncs to contain a nil Fn field.
+			// We will validate registered fieldfuncs later in getType.
+			if fieldFuncKind := reflect.ValueOf(method.Fn).Kind(); fieldFuncKind != reflect.Func {
+				continue
+			}
+
+			fieldFuncType := reflect.TypeOf(method.Fn)
+
+			for i := 0; i < fieldFuncType.NumIn(); i++ {
+				arg := fieldFuncType.In(i)
+				if arg == contextType {
+					continue
+				}
+
+				if err := checkTypeNameUniqueness(arg, typesToPackages); err != nil {
+					return nil, err
+				}
+			}
+
+			for i := 0; i < fieldFuncType.NumOut(); i++ {
+				ret := fieldFuncType.Out(i)
+				if ret == errType {
+					continue
+				}
+
+				if err := checkTypeNameUniqueness(ret, typesToPackages); err != nil {
+					return nil, err
+				}
+			}
+		}
+	}
+
+	for enum := range sb.enumMappings {
+		enumType := enum.Name()
+		enumPkg := enum.PkgPath()
+		if packageName, ok := typesToPackages[enumType]; ok && packageName != enumPkg {
+			return nil, fmt.Errorf(DuplicateTypeNameErrFormat, enumType, packageName, enumPkg)
+		}
+
+		typesToPackages[enumType] = enumPkg
```

Comment:
Readability: do you think move this entire block to a helper function will make the `Build` easier to read? Since the main functionality of the 'Build` is to build the schema rather than checking the uniqueness.

Reply:
Done!